plugins/hydra_optuna_sweeper/tests/test_optuna_sweeper_plugin.py

```diff
@@ -124,15 +116,13 @@ def test_launch_jobs(hydra_sweep_runner: TSweepRunner) -> None:
         assert sweep.returns is None
 
 
-# https://github.com/pyreadline/pyreadline/issues/65
-@mark.filterwarnings("ignore::DeprecationWarning")
 @mark.parametrize("with_commandline", (True, False))
 def test_optuna_example(with_commandline: bool, tmpdir: Path) -> None:
     cmd = [
         "example/sphere.py",
         "--multirun",
         "hydra.sweep.dir=" + str(tmpdir),
-        "hydra.sweeper.n_trials=20",
+        "hydra.sweeper.n_trials=100",
```

Comment:
I had to increase the `n_trials`, somehow having only 20 is no longer enough with 20 trials with the newer version optuna for 
```
tests/test_optuna_sweeper_plugin.py::test_optuna_example[False]
```
the test failed with
```
E       assert -1.0 == 0

tests/test_optuna_sweeper_plugin.py:149: AssertionError
```

Reply:
This may be because of [this change](https://github.com/optuna/optuna/commit/5db7b5bc0b1442d5429b7ba816e4f4271ef94b94#diff-98431f6c0563769a069980db67d10bde2830661626057a8e69ba3e46f8e98ed4) of `RandomSampler` . We extracted the search space transformation (i.e., transformation to unit cube), and shared it with the bult-in samplers.

Previously, we used `numpy.random.RandomState.randint` to sample integer and categorical values, but the current version uses `numpy.random.RandomState.uniform`. So, the suggested values will be changed.

In this test, we simply search the solution randomly, and I guess the combination is current random seed (123) and the `uniform` method showed worse performance by chance.

Reply:
How long does the test take now?

Reply:
with 100 trials
```
================================================================================================================================ slowest 4 durations ================================================================================================================================
14.22s call     plugins/hydra_optuna_sweeper/tests/test_optuna_sweeper_plugin.py::test_optuna_example[False]
12.33s call     plugins/hydra_optuna_sweeper/tests/test_optuna_sweeper_plugin.py::test_optuna_example[True]
```

before this change
```
4.83s call     plugins/hydra_optuna_sweeper/tests/test_optuna_sweeper_plugin.py::test_optuna_example[True]
3.76s call     plugins/hydra_optuna_sweeper/tests/test_optuna_sweeper_plugin.py::test_optuna_example[False]
```

@toshihikoyanase thanks for taking a look! that makes sense. Any recommendation on how do we change the test (or maybe likely the example app) to speed the tests up a bit?

Reply:
I'm sorry for the delayed response.

> Any recommendation on how do we change the test (or maybe likely the example app) to speed the tests up a bit?

I may not answer your question directly, but I'd like to change the scenario of the test case.
I roughly calculated the probabilities to pass the test cases, and I found that the current code is a bit fragile to changes in Optuna's implementation.

The `sphere.py` has 69 possible values (23 for `x` and 3 for `y`), and the `RandomSampler` will find the best solution with the following probabilities (`1 - ((69 - 1)/ 69) ** N_TRIALS`).

| Trials | Probability to find the best |
| --- | --- |
| 20 | 0.253 |
| 40 |  0.442 |
| 100 | 0.767 |
| 1000 | 0.999 |

So, we need 1000 trials to keep the test robust to Optuna's changes.

Instead of finding the best solution in the unit test, I think we can confirm the plugin code by comparing the results returned by this plugin with those in Optuna storage. This scenario seems straightforward since we can ensure that the results of plugin are the same as the results of Optuna.

The scenario is as follows:

1. Set Oputna storage and study name to the command-line options to persist the optimization results
2. Load the study from the Optuna storage
3. Fetch the best solution in the Optuna storage and compare it with plugin's result.

```python
@mark.parametrize("with_commandline", (True, False))
def test_optuna_example(with_commandline: bool, tmpdir: Path) -> None:
    storage = "sqlite:///" + os.path.join(str(tmpdir), "test.db")
    study_name = "test-optuna-example"
    cmd = [
        "example/sphere.py",
        "--multirun",
        "hydra.sweep.dir=" + str(tmpdir),
        "hydra.sweeper.n_trials=20",
        "hydra.sweeper.n_jobs=1",  # This is the out-of-scope, but `n_jobs` should be one for reproducibility.
        "hydra.sweeper.storage=" + storage,  # 1. Add Optuna storage
        "hydra.sweeper.study_name=" + study_name,  # 1. Add study name
        "hydra/sweeper/sampler=random",
        "hydra.sweeper.sampler.seed=123",
    ]
    if with_commandline:
        cmd += [
            "x=choice(0, 1, 2)",
            "y=0",  # Fixed parameter
        ]
    run_python_script(cmd)
    returns = OmegaConf.load(f"{tmpdir}/optimization_results.yaml")
    # 2. Load the study to retrieve the results in the Optuna's storage.
    study = optuna.load_study(storage=storage, study_name=study_name)
    best_trial = study.best_trial
    assert isinstance(returns, DictConfig)
    assert returns.name == "optuna"
    assert returns["best_params"]["x"] == best_trial.params["x"]  # 3. Compare plugin's result with Optuna's one.
    if with_commandline:
        assert "y" not in returns["best_params"]
    else:
        assert returns["best_params"]["y"] == best_trial.params["y"]  # 3. Compare plugin's result with Optuna's one.
    assert returns["best_value"] == best_trial.value  # 3. Compare plugin's result with Optuna's one.
```

The execution times are as follows:

The new scenario
```
3.26s call     plugins/hydra_optuna_sweeper/tests/test_optuna_sweeper_plugin.py::test_optuna_example[False]
3.00s call     plugins/hydra_optuna_sweeper/tests/test_optuna_sweeper_plugin.py::test_optuna_example[True]
```

This PR (with 100 trials)
```
8.89s call     plugins/hydra_optuna_sweeper/tests/test_optuna_sweeper_plugin.py::test_optuna_example[False]
8.79s call     plugins/hydra_optuna_sweeper/tests/test_optuna_sweeper_plugin.py::test_optuna_example[True]
```

master
```
2.38s call     plugins/hydra_optuna_sweeper/tests/test_optuna_sweeper_plugin.py::test_optuna_example[False]
2.32s call     plugins/hydra_optuna_sweeper/tests/test_optuna_sweeper_plugin.py::test_optuna_example[True]
```


What do you think?
If the change of the test scenario is out-of-scope of this PR, I'll create a new PR.

Reply:
Thanks @toshihikoyanase. This sounds good to me.
We are currently heads down in internal Hydra migration at Facebook. Can you take over this small effort and come up with a diff upgrade to the latest Optuna and fixes the tests as you are suggesting at the same time?

Reply:
@omry Sure! I'll take over this task.
Should I send a new PR or just add my commits to this PR?